// draft-js-resizeable-plugin/lib/decorators/resizeable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var resizeableRatioUtil = function resizeableRatioUtil(ratio, padding) {
  return {
    ratioContainerStyle: {
      position: 'relative'
    }, ratioContentStyle: {
      position: 'absolute',
      top: padding ? '-' + padding + 'px' : 0,
      left: padding ? '-' + padding + 'px' : 0,
      bottom: 0,
      right: 0,
      padding: padding ? padding + 'px' : 0
    }, createRatioPlaceholder: function createRatioPlaceholder() {
      return _react2.default.createElement('div', { style: { display: 'block', width: '100%', paddingTop: ratio * 100 + '%' } });
    }
  };
};

// Get a component's display name
var getDisplayName = function getDisplayName(WrappedComponent) {
  var component = WrappedComponent.WrappedComponent || WrappedComponent;
  return component.displayName || component.name || 'Component';
};

function round(x, steps) {
  return Math.ceil(x / steps) * steps;
}

// Export

exports.default = function (options) {
  return function (WrappedComponent) {
    var _class, _temp2;

    return _temp2 = _class = function (_Component) {
      _inherits(BlockResizeableDecorator, _Component);

      function BlockResizeableDecorator() {
        var _Object$getPrototypeO;

        var _temp, _this, _ret;

        _classCallCheck(this, BlockResizeableDecorator);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_Object$getPrototypeO = Object.getPrototypeOf(BlockResizeableDecorator)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.state = {
          hoverPosition: {},
          clicked: false
        }, _this.setEntityData = function (data) {
          var setEntityData = _this.props.blockProps.setEntityData;

          setEntityData(data);
        }, _this.mouseLeave = function () {
          if (!_this.state.clicked) {
            _this.setState({ hoverPosition: {} });
          }
        }, _this.mouseMove = function (e) {
          var _this$props = _this.props;
          var vertical = _this$props.vertical;
          var horizontal = _this$props.horizontal;


          var hoverPosition = _this.state.hoverPosition;
          var tolerance = 6;
          var pane = _this.DOMNode;

          var b = pane.getBoundingClientRect();
          var x = e.clientX - b.left;
          var y = e.clientY - b.top;

          var isTop = vertical && vertical !== 'auto' ? y < tolerance : false;
          var isLeft = horizontal ? x < tolerance : false;
          var isRight = horizontal ? x >= b.width - tolerance : false;
          var isBottom = vertical && vertical !== 'auto' ? y >= b.height - tolerance && y < b.height : false;

          var canResize = isTop || isLeft || isRight || isBottom;

          var newHoverPosition = {
            isTop: isTop, isLeft: isLeft, isRight: isRight, isBottom: isBottom, canResize: canResize
          };

          if (Object.keys(newHoverPosition).filter(function (key) {
            return hoverPosition[key] !== newHoverPosition[key];
          }).length) {
            _this.setState({ hoverPosition: newHoverPosition });
          }
        }, _this.mouseDown = function (event) {
          // No mouse-hover-position data? Nothing to resize!
          if (!_this.state.hoverPosition.canResize) {
            return undefined;
          }
          var _this$props2 = _this.props;
          var resizeSteps = _this$props2.resizeSteps;
          var vertical = _this$props2.vertical;
          var horizontal = _this$props2.horizontal;
          var hoverPosition = _this.state.hoverPosition;
          var isTop = hoverPosition.isTop;
          var isLeft = hoverPosition.isLeft;
          var isRight = hoverPosition.isRight;
          var isBottom = hoverPosition.isBottom;


          var component = _this.DOMNode;
          var startX = event.clientX;
          var startY = event.clientY;
          var startWidth = parseInt(document.defaultView.getComputedStyle(component).width, 10);
          var startHeight = parseInt(document.defaultView.getComputedStyle(component).height, 10);

          // Do the actual drag operation
          var doDrag = function doDrag(dragEvent) {
            var width = startWidth + dragEvent.clientX - startX;
            var height = startHeight + dragEvent.clientY - startY;
            var b = component.parentElement.parentElement;
            width = b.clientWidth < width ? b.clientWidth : width;
            height = b.clientHeight < height ? b.clientHeight : height;

            var widthPerc = 100 / b.clientWidth * width;
            var heightPerc = 100 / b.clientHeight * height;

            var newState = {};
            if ((isLeft || isRight) && horizontal === 'relative') {
              newState.width = resizeSteps ? round(widthPerc, resizeSteps) : widthPerc;
            } else if ((isLeft || isRight) && horizontal === 'absolute') {
              newState.width = resizeSteps ? round(width, resizeSteps) : width;
            }

            if ((isTop || isBottom) && vertical === 'relative') {
              newState.height = resizeSteps ? round(heightPerc, resizeSteps) : heightPerc;
            } else if ((isTop || isBottom) && vertical === 'absolute') {
              newState.height = resizeSteps ? round(height, resizeSteps) : height;
            }

            _this.setState(newState);
            dragEvent.stopPropagation();
            dragEvent.preventDefault();
            return false;
          };

          // Finished dragging
          var stopDrag = function stopDrag(e) {
            document.removeEventListener('mousemove', doDrag, false);
            document.removeEventListener('mouseup', stopDrag, false);

            var _this$state = _this.state;
            var width = _this$state.width;
            var height = _this$state.height;

            _this.setState({ clicked: false });
            setTimeout(function () {
              _this.setEntityData({ width: width, height: height });
            });

            e.stopPropagation();
            return false;
          };

          document.addEventListener('mousemove', doDrag, false);
          document.addEventListener('mouseup', stopDrag, false);

          _this.setState({ clicked: true });
          event.stopPropagation();
          event.preventDefault();
          return false;
        }, _temp), _possibleConstructorReturn(_this, _ret);
      }
      // Statics


      _createClass(BlockResizeableDecorator, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
          this.componentDidUpdate();
        }
      }, {
        key: 'componentWillUpdate',
        value: function componentWillUpdate() {
          if (this.DOMNode) {
            this.DOMNode.removeEventListener('mouseleave', this.mouseLeave);
            this.DOMNode.removeEventListener('mousemove', this.mouseMove);
            this.DOMNode.removeEventListener('mousedown', this.mouseDown);
          }
        }
      }, {
        key: 'componentDidUpdate',
        value: function componentDidUpdate() {
          this.DOMNode = _reactDom2.default.findDOMNode(this);
          var readOnly = this.props.blockProps.pluginEditor.getReadOnly();

          if (!readOnly && this.DOMNode) {
            this.DOMNode.addEventListener('mouseleave', this.mouseLeave);
            this.DOMNode.addEventListener('mousemove', this.mouseMove);
            this.DOMNode.addEventListener('mousedown', this.mouseDown);
          }
        }

        // Handle mousedown for resizing

      }, {
        key: 'render',
        value: function render() {
          var _props = this.props;
          var blockProps = _props.blockProps;
          var vertical = _props.vertical;
          var horizontal = _props.horizontal;
          var ratio = _props.ratio;
          var style = _props.style;
          var _state = this.state;
          var width = _state.width;
          var height = _state.height;
          var hoverPosition = _state.hoverPosition;
          var isTop = hoverPosition.isTop;
          var isLeft = hoverPosition.isLeft;
          var isRight = hoverPosition.isRight;
          var isBottom = hoverPosition.isBottom;


          var styles = _extends({ position: 'relative' }, style);

          if (horizontal === 'auto') {
            styles.width = 'auto';
          } else if (horizontal === 'relative') {
            styles.width = (width || blockProps.entityData.width || 40) + '%';
          } else if (horizontal === 'absolute') {
            styles.width = styles.width = (width || blockProps.entityData.width || 40) + 'px';
          }

          if (vertical === 'auto') {
            styles.height = 'auto';
          } else if (vertical === 'relative') {
            styles.height = (height || blockProps.entityData.height || 40) + '%';
          } else if (vertical === 'absolute') {
            styles.height = (height || blockProps.entityData.height || 40) + 'px';
          }

          // Handle cursor
          if (isRight && isBottom || isLeft && isTop) {
            styles.cursor = 'nwse-resize';
          } else if (isRight && isTop || isBottom && isLeft) {
            styles.cursor = 'nesw-resize';
          } else if (isRight || isLeft) {
            styles.cursor = 'ew-resize';
          } else if (isBottom || isTop) {
            styles.cursor = 'ns-resize';
          } else {
            styles.cursor = 'default';
          }

          if (ratio) {
            return _react2.default.createElement(WrappedComponent, _extends({}, this.props, { style: styles }, resizeableRatioUtil(ratio, 3)));
          }

          return _react2.default.createElement(WrappedComponent, _extends({}, this.props, { style: styles }));
        }
      }]);

      return BlockResizeableDecorator;
    }(_react.Component), _class.displayName = 'BlockDraggable(' + getDisplayName(WrappedComponent) + ')', _class.WrappedComponent = WrappedComponent.WrappedComponent || WrappedComponent, _class.defaultProps = _extends({
      horizontal: 'relative',
      vertical: false,
      ratio: null,
      resizeSteps: 5,
      handles: false,
      caption: false
    }, options), _temp2;
  };
};
